### PR TITLE
[proto] Added functional rotate_bounding_box op

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to #5514

## Description:
- Added functional `rotate_bounding_box` op
- Added warning if `expand=True` and `center is not None` and disabled provided `center`
- Added tests

### Results on synthetic images/bboxes:

<details>
<summary>
Code
</summary>

```python
import numpy as np

import torch
import torchvision
from torchvision.prototype import features
from torchvision.prototype.transforms.functional import rotate_bounding_box, rotate_image_tensor


size = (64, 72)
# xyxy format
in_boxes = [
    [10, 15, 25, 35],
    [50, 5, 70, 22],
    [45, 46, 56, 62],
]

im1 = 255 * np.ones(size + (3, ), dtype=np.uint8)
for in_box in in_boxes:
    im1[in_box[1]:in_box[3], in_box[0]:in_box[2], :] = (127, 127, 127)
    
t_im1 = torch.tensor(im1).permute(2, 0, 1).view(1, 3, *size)

in_boxes = features.BoundingBox(
    in_boxes, format=features.BoundingBoxFormat.XYXY, image_size=size
)
    
angle = 22
center = None
expand = True

out_boxes = rotate_bounding_box(
    in_boxes, 
    in_boxes.format,
    in_boxes.image_size,
    angle,
    expand=expand,
    center=center
)
print(out_boxes)

t_im2 = rotate_image_tensor(t_im1, angle, expand=expand, center=center)


import cv2
import matplotlib.pyplot as plt
%matplotlib inline


plt.figure(figsize=(14, 10))

plt.subplot(1,2,1)
plt.title("Input image + bboxes")
r1 = t_im1[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for in_box in in_boxes:    
    r1 = cv2.rectangle(r1, (in_box[0].item(), in_box[1].item()), (in_box[2].item(), in_box[3].item()), (255, 127, 0))
plt.imshow(r1)


plt.subplot(1,2,2)
plt.title("Output image + bboxes")
r2 = t_im2[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for out_box in out_boxes:
    out_box = np.round(out_box.cpu().numpy()).astype("int32")
    r2 = cv2.rectangle(r2, (out_box[0], out_box[1]), (out_box[2], out_box[3]), (255, 127, 0), 0)
plt.imshow(r2)
```
</details>

image

### Compare to [D2](https://github.com/facebookresearch/detectron2/blob/9258799e4e72786edd67940872e0ed2c4387aac5/detectron2/data/transforms/transform.py#L162):

TL;DR: matches the behaviour if expand=False and ~1 pixel difference otherwise.
This is due to cast to int : https://github.com/facebookresearch/detectron2/commit/63ea9001838bfa1078c60a878db9caf6c4246adb#diff-98378cbc4207da5ac829ceff03d62c48abbfc564911ac7d64c2f1d137b4f3474R130 and the way how affine matrix is computed : https://github.com/facebookresearch/detectron2/commit/63ea9001838bfa1078c60a878db9caf6c4246adb#diff-98378cbc4207da5ac829ceff03d62c48abbfc564911ac7d64c2f1d137b4f3474R171 

<details>
<summary>
Code
</summary>

```python
import numpy as np


size = image_size = (64, 64)
in_boxes = [
    [1, 1, 5, 5],
    [1, image_size[0] - 6, 5, image_size[0] - 2],
    [image_size[1] - 6, image_size[0] - 6, image_size[1] - 2, image_size[0] - 2],    
    [image_size[1] // 2 - 10, image_size[0] // 2 - 10, image_size[1] // 2 + 10, image_size[0] // 2 + 10],
]

im1 = 255 * np.ones(size + (3, ), dtype=np.uint8)
for in_box in in_boxes:
    im1[in_box[1]:in_box[3], in_box[0]:in_box[2], :] = (127, 127, 127)


angle = 34
expand = False
center = None if expand else [12, 23]

from detectron2.data.transforms import RotationTransform, AugmentationList
from detectron2.data.transforms import AugInput
import cv2

inpt = AugInput(im1, boxes=np.array(in_boxes, dtype="float32"))

augs = AugmentationList([RotationTransform(*size, angle, expand=expand, center=center, interp=cv2.INTER_NEAREST), ])
out = augs(inpt)
print(inpt.boxes)

import torch
import torchvision
from torchvision.prototype import features
from torchvision.prototype.transforms.functional import rotate_bounding_box, rotate_image_tensor


t_im1 = torch.tensor(im1).permute(2, 0, 1).view(1, 3, *size)

in_boxes = features.BoundingBox(
    in_boxes, format=features.BoundingBoxFormat.XYXY, image_size=size
)
    

out_boxes = rotate_bounding_box(
    in_boxes, 
    in_boxes.format,
    in_boxes.image_size,
    angle,
    expand=expand,
    center=center
)

from torchvision.prototype.transforms.functional import rotate_image_tensor

t_im2 = rotate_image_tensor(t_im1, angle, expand=expand, center=center)

import cv2
import matplotlib.pyplot as plt
%matplotlib inline


plt.figure(figsize=(14, 8))
plt.subplot(1,2,1)
plt.title("D2 input: image + bboxes")
r1 = im1
for in_box in in_boxes:    
    r1 = cv2.rectangle(r1, (in_box[0].item(), in_box[1].item()), (in_box[2].item(), in_box[3].item()), (255, 127, 0))
plt.imshow(r1)

ax = plt.subplot(1,2,2)
plt.title(f"D2 output: image + bboxes, angle={angle}, expand={expand}")
r2 = inpt.image
for out_box in inpt.boxes:
    out_box = np.round(out_box).astype("int32")
    r2 = cv2.rectangle(r2, (out_box[0], out_box[1]), (out_box[2], out_box[3]), (255, 127, 0), 1)
plt.imshow(r2)

plt.figure(figsize=(14, 8))

ax = plt.subplot(1,2,1)
plt.title("TV input: image + bboxes")
r1 = t_im1[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for in_box in in_boxes:    
    r1 = cv2.rectangle(r1, (in_box[0].item(), in_box[1].item()), (in_box[2].item(), in_box[3].item()), (255, 127, 0))
plt.imshow(r1)


ax = plt.subplot(1,2,2)
plt.title(f"TV output: image + bboxes, angle={angle}, expand={expand}")
r2 = t_im2[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for out_box in out_boxes:
    out_box = np.round(out_box.cpu().numpy()).astype("int32")
    r2 = cv2.rectangle(r2, (out_box[0], out_box[1]), (out_box[2], out_box[3]), (255, 127, 0), 0)
plt.imshow(r2)
```
</details>

### CODE REVIEW

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 1]
```diff
@@ -198,24 +199,28 @@ def affine_image_pil(
     return _FP.affine(img, matrix, interpolation=pil_modes_mapping[interpolation], fill=fill)
 
 
-def affine_bounding_box(
+def _affine_bounding_box_xyxy(
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_geometry.py:202]: Nit: Why add the implementation in a private method? As far as I understand, rotate could have used the public affine directly. Am I missing something?

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 2]
```diff
@@ -259,9 +305,12 @@ def rotate_image_tensor(
 ) -> torch.Tensor:
     center_f = [0.0, 0.0]
     if center is not None:
-        _, height, width = get_dimensions_image_tensor(img)
-        # Center values should be in pixel coordinates but translated such that (0, 0) corresponds to image center.
-        center_f = [1.0 * (c - s * 0.5) for c, s in zip(center, [width, height])]
+        if expand:
+            warnings.warn("If provided center argument is ignored if expand is True")
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_geometry.py:309]:
```suggestion
            warnings.warn("The provided center argument is ignored if expand is True")
```

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 3]
```diff
@@ -277,11 +326,35 @@ def rotate_image_pil(
     fill: Optional[List[float]] = None,
     center: Optional[List[float]] = None,
 ) -> PIL.Image.Image:
+    if center is not None and expand:
+        warnings.warn("If provided center argument is ignored if expand is True")
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_geometry.py:330]:
```suggestion
        warnings.warn("The provided center argument is ignored if expand is True")
```

**Review by datumbox — APPROVED** [review]
Thanks! LGTM with a couple of non-blocking nits for your consideration.

File: torchvision/prototype/transforms/functional/_geometry.py (hunk 1 above)
Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:202]: It is a refactoring of `affine_bounding_box` and this way I also avoid to expose `expand` argument that is not present for `affine_bounding_box`.

Comment (datumbox) [reply torchvision/prototype/transforms/functional/_geometry.py:201]: Thanks for the clarification! So because `affine_image_tensor` doesn't expose an expand, you propose not to expose it on `affine_bounding_box`; makes sense. Should we be exposing expand to affine as an optional parameter or you think that's unnecessary?

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:201]: I know that albu has `expand` as arg for `affine` transformation. I do not know if there is a strong requirement for that. Another point is that we have to code that for `affine_image_tensor` as there is no such option currently. We can create an RFC for that and see community feedback ?

Comment (datumbox) [reply torchvision/prototype/transforms/functional/_geometry.py:201]: Sounds good. Let's create an RFC and move on. We can always revise and add it later.
